zio/zngio/reader.go

```diff
@@ -170,17 +176,75 @@ again:
 	return rec, nil, nil
 }
 
-func (r *Reader) readUvarint() (int, error) {
-	b, err := r.peeker.Peek(binary.MaxVarintLen64)
-	if err != nil && err != io.EOF && err != peeker.ErrTruncated {
-		return 0, zng.ErrBadFormat
+// read returns an error if fewer than n bytes are available.
+func (r *Reader) read(n int) ([]byte, error) {
+	if r.uncompressed != nil {
+		if n > MaxSize {
+			return nil, errors.New("zngio: read exceeds MaxSize")
+		}
+		buf := r.uncompressed.Next(n)
+		if len(buf) < n {
+			return nil, errors.New("zngio: short read")
+		}
+		if r.uncompressed.Len() == 0 {
+			r.uncompressed = nil
+		}
+		return buf, nil
+	}
+	b, err := r.peeker.Read(n)
+	r.peekerOffset += int64(len(b))
+	return b, err
+}
+
+func (r *Reader) readCompressed() error {
+	format, err := r.readUvarint()
+	if err != nil {
+		return err
+	}
+	if zng.CompressionFormat(format) != zng.CompressionFormatLZ4 {
+		return fmt.Errorf("zngio: unknown compression format 0x%x", format)
+	}
+	uncompressedLen, err := r.readUvarint()
+	if err != nil {
+		return err
+	}
+	if uncompressedLen > MaxSize {
+		return errors.New("zngio: uncompressed length exceeds MaxSize")
+	}
+	compressedLen, err := r.readUvarint()
+	if err != nil {
+		return err
 	}
-	v, n := binary.Uvarint(b)
-	if n <= 0 {
-		return 0, zng.ErrBadFormat
+	zbuf, err := r.read(compressedLen)
+	if err != nil {
+		return err
+	}
+	if cap(r.uncompressedBuf) < uncompressedLen {
+		r.uncompressedBuf = make([]byte, uncompressedLen)
+	}
+	ubuf := r.uncompressedBuf[:uncompressedLen]
+	n, err := lz4.UncompressBlock(zbuf, ubuf)
+	if err != nil {
+		return fmt.Errorf("zngio: %w", err)
+	}
+	if n != uncompressedLen {
+		return fmt.Errorf("zngio: got %d uncompressed bytes, expected %d", n, uncompressedLen)
+	}
+	r.uncompressed = bytes.NewBuffer(ubuf)
+	return nil
+}
+
+func (r *Reader) readUvarint() (int, error) {
+	u64, err := binary.ReadUvarint(r)
+	return int(u64), err
+}
+
+func (r *Reader) ReadByte() (byte, error) {
```

Comment:
I think this should be a method on Peeker.  Useful to be there and more efficient.

Reply:
@mccanne: How does a9e91b854e9716fe56f68800761eea32d0fd4f3a look for that?